### PR TITLE
Update for go-ipfs-files refactor

### PR DESCRIPTION
**Pull request by magik6k**
Depends on ipfs/go-ipfs-files#2

### CODE REVIEW

**Comment by magik6k** [comment]
Blocked by https://github.com/ipfs/go-ipfs/pull/5762 which is blocked by https://github.com/ipfs/go-ipfs/pull/5526

**Comment by Stebalien** [comment]
Ok, I'm just going to rebase #5526.

**Comment by Stebalien** [comment]
I'll rebase this once the files refactor settles.

File: core/coreapi/unixfile.go [hunk 1]
```diff
@@ -23,160 +17,115 @@ const prefetchFiles = 4
...
-type ufsFile struct {
-	uio.DagReader
-
-	name string
-	path string
-}
+func (d *ufsDirectory) Entries() files.DirIterator {
+	fileCh := make(chan *ipld.Link, prefetchFiles)
+	go func() {
+		d.dir.ForEachLink(d.ctx, func(link *ipld.Link) error {
```
Comment (Stebalien) [inline core/coreapi/unixfile.go]: What if this fails? I believe we need to stash the error.

Trivial case: context is canceled. In that case, `iterator.Err()` should return the context cancellation error.

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -240,14 +236,14 @@ func (i *gatewayHandler) getOrHeadHandler(ctx context.Context, w http.ResponseWr
 	// TODO: break this out when we split /ipfs /ipns routes.
 	modtime := time.Now()
 
-	if strings.HasPrefix(urlPath, ipfsPathPrefix) && !dir {
-		w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
+	if f, ok := dr.(files.File); ok {
+		if strings.HasPrefix(urlPath, ipfsPathPrefix) {
+			w.Header().Set("Cache-Control", "public, max-age=29030400, immutable")
```
Comment (Stebalien) [inline core/corehttp/gateway_handler.go]: Don't we want to set this for directories as well?

File: core/coreunix/add.go [hunk 3]
```diff
@@ -467,8 +464,8 @@ func (adder *Adder) addFile(file files.File) error {
 	}
 	adder.liveNodes++
 
-	if file.IsDirectory() {
-		return adder.addDir(file)
+	if dir, ok := file.(files.Directory); ok {
```
Comment (Stebalien) [inline core/coreunix/add.go]: Nit: while a larger edit, changing this function to a type switch would (IMO) make it easier to read.

File: core/coreunix/add.go [hunk 4]
```diff
@@ -534,25 +533,23 @@ func (adder *Adder) addDir(dir files.File) error {
...
+		err = adder.addFile(fpath, it.Node())
 		if err != nil {
 			return err
 		}
 	}
+	if it.Err() != nil {
+		return it.Err()
+	}
 
 	return nil
```
Comment (Stebalien) [inline core/coreunix/add.go]: We can just return `it.Err()` here (slightly simpler).

File: core/commands/tar.go [hunk 5]
```diff
@@ -44,21 +45,23 @@ represent it.
 			return err
 		}
 
-		fi, err := req.Files.NextFile()
-		if err != nil {
-			return err
+		it := req.Files.Entries()
+		if !it.Next() && it.Err() != nil {
+			return it.Err()
```
Comment (Stebalien) [inline core/commands/tar.go]: here

File: core/commands/tar.go [hunk 6]
```diff
@@ -44,21 +45,23 @@ represent it.
...
+		}
+		if files.FileFromEntry(it) == nil {
```
Comment (Stebalien) [inline core/commands/tar.go]: here

File: core/commands/object/patch.go [hunk 7]
```diff
@@ -107,12 +111,15 @@ Example:
 			return err
 		}
 
-		data, err := req.Files.NextFile()
-		if err != nil {
-			return err
+		it := req.Files.Entries()
+		if !it.Next() && it.Err() != nil {
```
Comment (Stebalien) [inline core/commands/object/patch.go]: here

File: core/commands/object/patch.go [hunk 8]
```diff
@@ -107,12 +111,15 @@ Example:
...
+			return it.Err()
+		}
+		if files.FileFromEntry(it) == nil {
```
Comment (Stebalien) [inline core/commands/object/patch.go]: here

File: core/commands/object/patch.go [hunk 9]
```diff
@@ -60,12 +61,15 @@ the limit will not be respected by the network.
 			return err
 		}
 
-		data, err := req.Files.NextFile()
-		if err != nil {
-			return err
+		it := req.Files.Entries()
+		if !it.Next() && it.Err() != nil {
```
Comment (Stebalien) [inline core/commands/object/patch.go]: here

File: core/commands/object/patch.go [hunk 10]
```diff
@@ -60,12 +61,15 @@ the limit will not be respected by the network.
...
+			return it.Err()
+		}
+		if files.FileFromEntry(it) == nil {
```
Comment (Stebalien) [inline core/commands/object/patch.go]: here

File: core/commands/object/object.go [hunk 11]
```diff
@@ -391,9 +392,12 @@ And then run:
 			return err
 		}
 
-		input, err := req.Files.NextFile()
-		if err != nil && err != io.EOF {
-			return err
+		it := req.Files.Entries()
+		if !it.Next() && it.Err() != nil {
```
Comment (Stebalien) [inline core/commands/object/object.go]: here

File: core/commands/object/object.go [hunk 12]
```diff
@@ -391,9 +392,12 @@ And then run:
...
+			return it.Err()
+		}
+		if files.FileFromEntry(it) == nil {
```
Comment (Stebalien) [inline core/commands/object/object.go]: here

File: core/corehttp/gateway_handler.go (hunk 2 above)
Comment (magik6k) [reply core/corehttp/gateway_handler.go]: We don't as it would break with updates to dir-index-html

Comment (Stebalien) [reply core/corehttp/gateway_handler.go]: Yeah, I guess.

File: core/commands/cat.go [hunk 13]
```diff
@@ -122,12 +123,13 @@ func cat(ctx context.Context, api iface.CoreAPI, paths []string, offset int64, m
 			return nil, 0, err
 		}
 
-		file, err := api.Unixfs().Get(ctx, fpath)
+		f, err := api.Unixfs().Get(ctx, fpath)
 		if err != nil {
 			return nil, 0, err
 		}
 
-		if file.IsDirectory() {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply core/commands/cat.go]: I don't think we have proper support for them currently anyways. Changed the error to make it less confusing

**Comment by magik6k** [comment]
@Stebalien rebased, is there anything else blocking this or is it RFM?

**Pull request merged**